spanner/jdbc/src/main/java/com/example/spanner/jdbc/LoadCsvExample.java

```diff
@@ -0,0 +1,283 @@
+            case TIMESTAMP:
+              builder.set(columnName).to(com.google.cloud.Timestamp.parseTimestamp(recordValue));
+              break;
+            default:
+              System.out.print("Invalid Type. This type is not supported.");
+          }
+        }
+      }
+      mutations.add(builder.build());
+    }
+    CloudSpannerJdbcConnection spannerConnection = connection
+        .unwrap(CloudSpannerJdbcConnection.class);
+    spannerConnection.write(mutations);
+    spannerConnection.close();
+    System.out.println("Data successfully written into table.");
+  }
+
+  static void loadCsv() throws Exception {
+    // TODO(developer): Replace these variables before running the sample.
+    String projectId = "my-project-id";;
+    String instanceId = "my-instance-id";
+    String databaseId = "my-database-id";
+    String tableName = "my-table-name";
+    String filePath = "my-file-path";
+    String[] optFlags = {"my-opt-flag", "my-opt-arg"};
+    loadCsv(projectId, instanceId, databaseId, tableName,
+        filePath, optFlags);
+  }
+
+  static void loadCsv(String projectId, String instanceId, String databaseId,
+      String tableName, String filePath, String[] optFlags) throws Exception {
+
+    // Initialize option flags
+    Options opt = new Options();
+    opt.addOption("h", true, "File Contains Header");
+    opt.addOption("f", true, "Format Type of Input File "
+        + "(EXCEL, POSTGRESQL_CSV, POSTGRESQL_TEXT, DEFAULT)");
+    opt.addOption("n", true, "String Representing Null Value");
+    opt.addOption("d", true, "Character Separating Columns");
+    opt.addOption("e", true, "Character To Escape");
+    CommandLineParser clParser = new DefaultParser();
+    CommandLine cmd = clParser.parse(opt, optFlags);
+
+    SpannerOptions options = SpannerOptions.newBuilder().build();
+    Spanner spanner = options.getService();
+
+    // Initialize connection to Cloud Spanner
+    try (
+        Connection conn = DriverManager.getConnection(
+            String.format("jdbc:cloudspanner:/projects/%s/instances/%s/databases/%s",
+            projectId, instanceId, databaseId));
+        Reader in = new FileReader(filePath)
+    ) {
+
+      connection = conn;
+      parseTableColumns(tableName);
+      CSVFormat parseFormat = setFormat(cmd);
+
+      try (CSVParser parser = CSVParser.parse(in, parseFormat)) {
+        // If file has header, verify that header fields are valid
+        if (hasHeader && !isValidHeader(parser)) {
+          return;
+        }
+        // Write CSV record data to Cloud Spanner
+        try {
+          writeToSpanner(parser, tableName);
+        } catch (SQLException e) {
+          /* SQLExceptions are thrown when the table name cannot be queried for in the database
+          or the connection established does not have permissions to write data into the table */
+          System.out.println(e.getMessage());
+        }
+      }
+    } finally {
+      spanner.close();
+      connection.close();
```

Comment:
If you use try with resources you shouldn't need to close this. It seems like it would be cleaner to move connection out of the try with resources

Reply:
I tried to simplify to 2 tries instead of 3, let me know if that seems cleaner based on my understanding of your comment?